File: Problem_1.py
#
# @lc app=leetcode id=146 lang=python3
#
# [146] LRU Cache
#

# @lc code=start
'''
Time Complexity - O(1) for put and get as we are using hashMap for storing keys and node
Space Complexity - O(Capacity) as we are maintaining a HashMap for storing the keys and node and size at max would be same as capacity

This code works on Leetcode.
'''
class Node: #create the LinkedList Node
    def __init__(self, key, val):
        self.key = key
        self.val = val
        self.next = None
        self.prev = None

class LRUCache:

    def __init__(self, capacity: int): #create LinkedList
        self.head = Node(-1,-1)
        self.tail = Node(-1,-1) 
        self.hashMap = {}
        self.capacity = capacity
        self.head.next = self.tail
        self.tail.prev = self.head

    def removeNode(self, node): #O(1). remove a node from the LinkedList
        node.next.prev = node.prev
        node.prev.next = node.next

    def addToHead(self, node): #O(1) Add a Node to head of the LinkedList
        node.next = self.head.next
        node.prev = self.head
        self.head.next = node 
        node.next.prev = node
        

    def get(self, key: int) -> int:
        if key not in self.hashMap: #check if key is present in the cache through HashMap
            return -1 #return -1 if not present
        node = self.hashMap[key] #else get the node as we have used it
        self.removeNode(node) #remove the node from the list and add put in the front as it becomes most recently used
        self.addToHead(node)
        return node.val  #returnt the value of the node
        

    def put(self, key: int, value: int) -> None:
        if key in self.hashMap: #if node present in cache, update value, take out from the LinkedList and place it at first
            node = self.hashMap[key]
            node.val = value
            self.removeNode(node)
            self.addToHead(node)
        else:
            if self.capacity == len(self.hashMap): #if node not present, check the capacity of the Cache,
                #remove the previous
                leastRecent = self.tail.prev #if current size equals capacity, remove the last node as its least recent
                self.removeNode(leastRecent)
                self.hashMap.pop(leastRecent.key)
            newNode = Node(key, value) #create a new node for new entry
            self.addToHead(newNode) #Add it to head as its most recently used
            self.hashMap[key] = newNode #establish entry in hashMap



# Your LRUCache object will be instantiated and called as such:
# obj = LRUCache(capacity)
# param_1 = obj.get(key)
# obj.put(key,value)
# @lc code=end

File: problem1.go
// Time Complexity : O(1)
// Space Complexity : O(capacity)
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this : No
//TC - O(1)
//SC  - O(capacity)

type LRUCache struct {
    capacity    int
    head        *Node
    tail        *Node
    hashmap     map[int]*Node     
}

type Node struct{
    key     int
    value   int
    next    *Node
    prev    *Node
}

func Constructor(capacity int) LRUCache {
    cache := LRUCache{
        capacity: capacity,
        hashmap: map[int]*Node{},
    }
    
    cache.head = &Node{
        key:    -1,
        value:  -1,
    }
    
    cache.tail = &Node{
        key:    -1,
        value:  -1,
    }
    
    cache.head.next = cache.tail
    cache.tail.prev = cache.head
    
    return cache
}


func (this *LRUCache)addToHead(node *Node){
    node.next = this.head.next
    node.prev = this.head
    this.head.next = node
    node.next.prev = node
}

func (this *LRUCache)removeNode(node *Node){
    node.prev.next = node.next
    node.next.prev = node.prev
}

func (this *LRUCache) Get(key int) int {
    node,ok := this.hashmap[key]
    
    if !ok{
        return -1
    }
    
    this.removeNode(node)
    this.addToHead(node)
    return node.value
}


func (this *LRUCache) Put(key int, value int)  {
    if node,ok := this.hashmap[key];ok{
        node.value = value
        this.removeNode(node)
        this.addToHead(node)
        
        return
    }
    
    
    if len(this.hashmap) == this.capacity{
        prev := this.tail.prev
        this.removeNode(prev)
        delete(this.hashmap,prev.key)
    }
    
    node:= &Node{
        key: key,
        value: value,
    }
    
    this.addToHead(node)
    this.hashmap[key] = node
    
}


/**
 * Your LRUCache object will be instantiated and called as such:
 * obj := Constructor(capacity);
 * param_1 := obj.Get(key);
 * obj.Put(key,value);
 */
